include/libsemigroups/to-todd-coxeter.hpp

```diff
@@ -15,62 +15,112 @@
+                         Result>,
+          Result>;
+#else
+  template <typename Result, typename Node>
+  auto to(congruence_kind knd, FroidurePinBase& fpb, WordGraph<Node> const& wg)
+      -> std::enable_if_t<
+          std::is_same_v<ToddCoxeter<typename Result::native_word_type>,
+                         Result>,
+          Result>;
+#endif
 
-    for (label_type a = 0; a < copy.out_degree(); ++a) {
-      copy.target_no_checks(0, a, fpb.position_of_generator_no_checks(a) + 1);
-    }
-
-    for (node_type n = 0; n < copy.number_of_nodes() - 1; ++n) {
-      for (label_type a = 0; a < copy.out_degree(); ++a) {
-        copy.target_no_checks(n + 1, a, wg.target_no_checks(n, a) + 1);
-      }
-    }
-    // TODO(1) move "copy" into ToddCoxeter.
-    return ToddCoxeter<Word>(knd, copy);
-  }
-
-  // TODO(0) allow template param "word_type" to be specified
-  template <typename Rewriter, typename ReductionOrder>
-  ToddCoxeter<word_type>
-  to_todd_coxeter(congruence_kind                                    knd,
-                  detail::KnuthBendixImpl<Rewriter, ReductionOrder>& kb) {
-    if (kb.number_of_classes() == POSITIVE_INFINITY) {
-      LIBSEMIGROUPS_EXCEPTION(
-          "cannot construct a ToddCoxeterImpl instance using the Cayley graph "
-          "of "
-          "an infinite KnuthBendixImpl<> object, maybe try ToddCoxeterImpl({}, "
-          "kb.presentation()) instead?",
-          kb.kind());
-    }
-    // TODO why are we doing this? Why not just use the active rules of kb?
-    auto fp = to_froidure_pin(kb);
-    return to_todd_coxeter(knd, fp, fp.right_cayley_graph());
-  }
+  //! \ingroup to_todd_coxeter_group
+  //!
+  //! \brief Convert a \ref_knuth_bendix object to a \ref_todd_coxeter object.
+  //!
+  //! This function converts the \ref_knuth_bendix object \p kb into a
+  //! \ref_todd_coxeter object using the right Cayley graph of the semigroup
+  //! represented by \p kb.
+  //!
+  //! \tparam Thing used for SFINAE, must be FroidurePin.
+  //! \tparam Word the type of the words used in relations in \p kb.
+  //! \tparam Rewriter the type of rewriter used by \p kb.
+  //! \tparam ReductionOrder the type of reduction ordering used by \p kb.
+  //!
+  //! \param knd the kind of the congruence being constructed.
+  //! \param kb the \ref_knuth_bendix object being converted.
+  //!
+  //! \returns A \ref_todd_coxeter object representing the trivial congruence
+  //! over the semigroup defined by \p kb.
+  //!
+  //! \throws LibsemigroupsException if `kb.kind()` is not \ref
+  //! congruence_kind::twosided.
+  //!
+  //! \throws LibsemigroupsException if `kb.number_of_classes()` is not finite.
+  //! Use ToddCoxeter(knd, kb.presentation()) in this case.
+  template <template <typename...> typename Thing,
+            typename Word,
+            typename Rewriter,
+            typename ReductionOrder>
+  auto to(congruence_kind knd, KnuthBendix<Word, Rewriter, ReductionOrder>& kb)
```

Comment:
Is there any reason why this is implemented for `KnuthBendix`, but the `to-froidure-pin` counterpart is implemented for `KnuthBendixImpl`?

Reply:
Yeah good question, we need the type of the words in the knuthbendix so we can use the same type in Todd-Coxeter